### PR TITLE
np.isnat

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Add support for `np.isnat`. Still a work in progress.

#4074

### CODE REVIEW

**Comment by esc** [comment]
@guilhermeleobas thanks for opening this PR. Your effort to improve Numba are appreciated! I have labeled this WIP for now, do ping us when it's ready for review.

**Comment by guilhermeleobas** [comment]
One test is failing due to big/little endian differences. As soon as I address those, this PR will be ready for a round of reviews.

**Comment by guilhermeleobas** [comment]
Ready for review as soon as CI passes

**Comment by esc** [comment]
@guilhermeleobas thanks, I have changed the status of this PR now.

File: numba/np/npyfuncs.py [hunk 1]
```diff
@@ -1473,6 +1473,13 @@ def np_datetime_isfinite_impl(context, builder, sig, args):
     return builder.icmp_unsigned('!=', args[0], npdatetime.NAT)
 
 
+def np_isnat_impl(context, builder, sig, args):
+    _check_arity_and_homogeneity(sig, args, 1, return_type=types.boolean)
+    return builder.or_(
+        builder.icmp_unsigned('==', args[0], npdatetime.NAT),
+        builder.icmp_unsigned('==', args[0], npdatetime.NAT_BIG_ENDIAN))
+
```
Comment (stuartarchibald) [inline numba/np/npyfuncs.py]: Are you sure this logic is valid? I think a big endian NaT would appear as a valid little endian datetime. e.g.
```python
from numba import njit
import numpy as np

@njit
def foo(a):
    return np.isnat(a)

BENaT = np.datetime64('NaT').byteswap()
arg = np.array(BENaT.view(np.int64), np.dtype('<M8[Y]'))
print("arg", arg)

print(foo.py_func(arg))
print(foo(arg))
```
on a LE machine produces:
```
arg 2098
False
True
```
I'd anticipate that the inverse would hold on a BE machine.

Comment (stuartarchibald) [reply numba/np/npyfuncs.py]: Further, it's probably worth checking if any of the other `is*` functions work on BE before going down the route of attempting to support it?

Comment (guilhermeleobas) [reply numba/np/npyfuncs.py]: What do you mean by _BE_ and _LE_?

Comment (guilhermeleobas) [reply numba/np/npyfuncs.py]: Hey @stuartarchibald, you're right. My bet is that the [current NumPy implementation](https://github.com/numpy/numpy/blob/0b9b176dcef7f14f56dbd543c89e6da6c6b7b2e4/numpy/core/src/umath/loops.c.src#L1085-L1092) for `np.isnat` does a byte-swapping of the underlying data. Thus, `np.isnat(np.array(['NaT'], dtype='>datetime64[Y]'))` is actually comparing `NPY_DATETIME_NAT == arg.byteswap()`

```python
>>> np.array(['NaT'], dtype='>datetime64[Y]').view(np.int64)
array([128])

>>> np.array(['NaT'], dtype='>datetime64[Y]').byteswap().view(np.int64)
array([-9223372036854775808])

>>> np.array(['NaT'], dtype='<datetime64[Y]').view(np.int64)
array([-9223372036854775808])
```

**Comment by guilhermeleobas** [comment]
Just notice that Numba already has a function to check if a value **isn't ** a NAT.
https://github.com/numba/numba/blob/13ece9b97e6f01f750e870347f231282325f60c3/numba/np/npdatetime.py#L92-L96

This will not work for the following input: `np.array(['NaT'], dtype='>datetime64[Y]')`

**Comment by stuartarchibald** [comment]
Thanks, fixes. Please could you add this to the docs here https://github.com/numba/numba/blame/e20b17900ca24bada7ce51accf656b5db39f8fc7/docs/source/reference/numpysupported.rst#L745 in a new section `Datetime functions`. Thanks.

**Comment by stuartarchibald** [comment]
> Thanks for the fixes. Please could you add this to the docs here https://github.com/numba/numba/blame/e20b17900ca24bada7ce51accf656b5db39f8fc7/docs/source/reference/numpysupported.rst#L745 in a new section `Datetime functions`. Thanks.

Patched in f5d250f

**Comment by guilhermeleobas** [comment]
Just saw the notification for this PR. Is there anything else to be done in this PR?

**Comment by stuartarchibald** [comment]
Assuming the docs build fine, all done, thanks for your efforts.

**Comment by guilhermeleobas** [comment]
Glad to help

**Comment by stuartarchibald** [comment]
Manually checked the doc build. Looks fine.

**Pull request merged**